functions.global.php

```diff
@@ -21,6 +21,108 @@
+		return;
+	}
+
+	// Look for when a function will be removed based on when it was deprecated.
+	$removed_version = jetpack_get_future_removed_version( $version );
+
+	// If we could find a version, let's log a message about when removal will happen.
+	if (
+		! empty( $removed_version )
+		&& ( defined( 'WP_DEBUG' ) && WP_DEBUG )
+	) {
+		error_log( // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+			sprintf(
+				/* Translators: 1. Function name. 2. Jetpack version number. */
+				__( 'The %1$s function will be removed from the Jetpack plugin in version %2$s.', 'jetpack' ),
+				$function,
+				$removed_version
+			)
+		);
+
+	}
+}
+add_action( 'deprecated_function_run', 'jetpack_deprecated_function', 10, 3 );
+
+/**
+ * Hook into Core's _deprecated_file
+ * Add more details about when a deprecated file will be removed.
+ *
+ * @since 8.8.0
+ *
+ * @param string $file        The file that was called.
+ * @param string $replacement The file that should have been included based on ABSPATH.
+ * @param string $version     The version of WordPress that deprecated the file.
+ * @param string $message     A message regarding the change.
+ */
+function jetpack_deprecated_file( $file, $replacement, $version, $message ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+	// Bail early for non-Jetpack deprecations.
+	if ( 0 !== strpos( $version, 'jetpack-' ) ) {
+		return;
+	}
+
+	// Look for when a file will be removed based on when it was deprecated.
+	$removed_version = jetpack_get_future_removed_version( $version );
+
+	// If we could find a version, let's log a message about when removal will happen.
+	if (
+		! empty( $removed_version )
+		&& ( defined( 'WP_DEBUG' ) && WP_DEBUG )
+	) {
+		error_log( // phpcs:ignore WordPress.PHP.DevelopmentFunctions.error_log_error_log
+			sprintf(
+				/* Translators: 1. File name. 2. Jetpack version number. */
+				__( 'The %1$s file will be removed from the Jetpack plugin in version %2$s.', 'jetpack' ),
+				$file,
+				$removed_version
+			)
+		);
+
+	}
+}
+add_action( 'deprecated_file_included', 'jetpack_deprecated_file', 10, 4 );
+
+/**
+ * Get the major version number of Jetpack 6 months after provided version.
+ * Useful to indicate when a deprecated function will be removed from Jetpack.
+ *
+ * @since 8.8.0
+ *
+ * @param string $version The version of WordPress that deprecated the function.
+ *
+ * @return string|float Return a Jetpack Major version number, or an empty string.
+ */
+function jetpack_get_future_removed_version( $version ) {
+	/*
+	 * Extract the version number from a deprecation notice.
+	 * (let's only keep the first decimal, e.g. 8.8 and not 8.8.0)
+	 */
+	preg_match( '#(([0-9]+\.[0-9]+)(?:\.[0-9]+)*)#', $version, $matches );
+
+	if ( ! empty( $matches[2] ) ) {
+		// We'll remove the function from the code 6 months later, thus 6 major versions later.
+		$removed_version = $matches[2] + 0.6;
```

Comment:
I like this approach too. We'd need to change it if we ever use semver (e.g. Jetpack 9.15), but I don't see us doing that anytime soon.

I haven't tried to break the regex, but should we cast matches as a float before adding .6 to it?

Reply:
I've made some changes to address this in 1a33e539b874fb98096c10b5c2b211e31ea56056